bgzf/reader.go

```diff
@@ -4,22 +4,81 @@ import (
 	"bufio"
 	"bytes"
 	"compress/gzip"
+	"io"
 	"log"
 	"math"
 	"os"
 )
 
-// Reader is the decompressor for BGZF files which is at it's core
+// BlockReader is the decompressor for BGZF files which is at it's core
 // a gzip reader. The BGZF reader retains a *os.File to use for
 // seek functionality as well as a *bufio.Reader to implement the
 // io.ByteReader interface, required for reading gzip files as
 // independent concatenated blocks, rather than a multistream.
-type Reader struct {
+type BlockReader struct {
 	file         *os.File
 	buf          *bufio.Reader
 	decompressor *gzip.Reader
 }
 
+type Reader struct {
+	br           *BlockReader
+	blk          *Block
+	intermediate bytes.Buffer
+	eof          bool
+}
+
+// Next returns the next n bytes from the bgzf block.
+// Next handles reading new block if more bytes are
+// requested than are in the currently store block.
+func (r *Reader) Next(n int) []byte {
+	// we have enough bytes in intermediate
+	if r.intermediate.Len() >= n {
+		return r.intermediate.Next(n)
+	}
+
+	// we need to go to block for extra bytes
+	if r.blk.Len() >= n-r.intermediate.Len() {
+		if r.intermediate.Len() == 0 {
+			return r.blk.Next(n)
+		} else {
+			blkLen := r.intermediate.Len()
+			return append(r.intermediate.Next(n), r.blk.Next(n-blkLen)...)
+		}
+	}
+
+	// not enough bytes in the currently stored block
+	_, err := r.intermediate.ReadFrom(r.blk)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	err = r.br.ReadBlock(r.blk)
+	if err == io.EOF {
+		r.eof = true
+	}
+
+	// return with fewest appends possible
+	if r.intermediate.Len() != 0 {
+		intLen := r.intermediate.Len()
+		return append(r.intermediate.Next(n), r.blk.Next(n-intLen)...)
+	}
+
+	return r.blk.Next(n)
+}
+
+// Read retrieves the next len(b) bytes from the reader
+// and automatically handles block reading.
+func (r *Reader) Read(b []byte) (n int, err error) {
+	retrieved := r.Next(len(b))
+	n = len(retrieved)
+	if r.eof {
+		err = io.EOF
+	}
+	copy(b, retrieved)
+	return
```

Comment:
confused by this syntax, should it be return n, err?

Reply:
This is a quirk of using named variables. The named variables are pre-allocated and may be set in the function. The empty return at the end returns whatever is the current value of n and err rather than explicitly stating `return n, err`.